gridftp/gsincftp/source/doc/man/ncftpget.1
.TH ncftpget 1 NcFTP Software
.SH NAME
ncftpget - Internet file transfer program for scripts
.SH "SYNOPSIS"
.PP
.B ncftpget
.RI [ "options" ]
.I "remote-host" "local-directory" "remote-files..."
.PP
.B ncftpget
-f login.cfg
.RI [ "options" ]
.I "local-directory" "remote-files..."
.PP
.B ncftpget
.RI [ "options" ]
.I "ftp://url.style/host/path/name"
.\"-------
.SH "OPTIONS"
.\"-------
.SS
Command line flags:
.TP 8
.BI "-u " "XX"
Use username
.I XX
instead of anonymous.
.TP 8
.BI "-p " "XX"
Use password
.I XX
with the username.
.TP 8
.BI "-P " "XX"
Use port number
.I XX
instead of the default FTP service port (21).
.TP 8
.BI "-j " "XX"
Use account
.I XX
in supplement to the username and password (deprecated).
.TP 8
.BI "-d " "XX"
Use the file
.I XX
for debug logging.
.TP 8
.B -a
Use ASCII transfer type instead of binary.
.TP 8
.BI "-t " "XX"
Timeout after
.I XX
seconds.
.TP 8
.BR "-v" "/" "-V"
Do (do not) use progress meters.
The default is to use progress meters if the output stream is a TTY.
.TP 8
.BI "-f " "XX"
Read the file
.I XX
for host, user, and password information.
.TP 8
.B -A
Append to local files, instead of overwriting them.
.TP 8
.BR "-z" "/" "-Z"
Do (do not) try to resume transfers.
The default is to try to resume (\-z).
.TP 8
.B -F
Use passive (PASV) data connections.
.TP 8
.B -DD
Delete remote file after successfully downloading it.
.TP 8
.B -R
Recursive mode; copy whole directory trees.
.TP 8
.BI "-r " "XX"
Redial a maximum of 
.I XX
times until connected to the remote FTP server.
.TP 8
.B -b
Run in background (by submitting a job to
.IR ncftpbatch ")."
.TP 8
.BI "-B " "XX"
Try setting the TCP/IP socket buffer size to
.I XX
bytes.
.\"-------
.SH "DESCRIPTION"
.\"-------
.PP
The
purpose of
.I ncftpget
is to do file transfers from the command-line
without entering an interactive shell.
This lets you write shell scripts or other unattended
processes that can do FTP.
It is also useful for advanced users who
want to retrieve files from the shell command line without
entering an interactive FTP program such as
.IR ncftp "."
.PP
One particularly useful feature of this program is that
you can give it a
.I uniform resource locator
as the only argument and the program will download
that file.
You can then copy and paste from your web browser
or newsreader and use that URL.
Example:
.RS 4
.sp
$ cd /tmp
.br
$ ncftpget ftp://ftp.probe.net/pub/ncftp/ncftp.tar.Z
.br
$ zcat ncftp.tar.Z | tar xf -
.br
.RE
.PP
By default the program tries to open the remote host
and login anonymously, but you can specify a username
and password information.
The
.B -u
option is used to specify the username to login as,
and the
.B -p
option is used to specify the password.
If you are running the program from the shell, you may
omit the
.B -p
option and the program will prompt you for the password.
.PP
Using the 
.B -u
and
.B -p
options are not recommended, because your account information
is exposed to anyone who can see your shell script or your
process information.  For example, someone using the
.I ps
program could see your password while the program runs.
.PP
You may use the
.B -f
option instead to specify a file with the account information.
However, this is still not secure because anyone who
has read access to the information file can see the
account information.
Nevertheless, if you choose to use the
.B -f
option the file should look something like this:
.RS 4
.sp
host Bozo.probe.net
.br
user gleason
.br
pass mypasswd
.br
.sp
.RE
Don't forget to change the permissions on this file
so no one else can read them.
.PP
The
.B -d
option is very useful when you are trying to diagnose
why a file transfer is failing.
It prints out the
entire FTP conversation to the file you specify, so
you can get an idea of what went wrong.  
If you specify the special name
.I stdout
as the name of the debugging output file, the output
will instead print to the screen.  Example:
.RS 4
.sp
$ ncftpget -d stdout ftp.probe.net . /pub/ncftp/README
.br
220: ftp.probe.net FTP server ready.
.br
Connected to ftp.probe.net.
.br
Cmd: USER anonymous
.br
331: Guest login ok, send your complete e-mail address as password.
.br
Cmd: PASS xxxxxxxx
.br
230: Welcome!
.br
Logged in to ftp.probe.net as anonymous.
.br
Cmd: TYPE I
.br
200: Type set to I.
.br
Cmd: SIZE /pub/ncftp/README
.br
550: /pub/ncftp/README: not a plain file.
.br
Cmd: MDTM /pub/ncftp/README
.br
550: /pub/ncftp/README: No such file or directory.
.br
Cmd: PORT 192,168,1,17,6,76
.br
200: PORT command successful.
.br
Cmd: RETR /pub/ncftp/README
.br
550: /pub/ncftp/README: No such file OR directory.
.br
ncftpget: file retrieval error: could not start data transfer.
.br
Cmd: QUIT
.br
221: Goodbye.
.br
.RE
.PP
Using ASCII mode is helpful when the text format of your host
differs from that of the remote host.
For example, if you are retrieving a \.TXT file from
a Windows-based host to a UNIX system, you could use the
.B -a
flag which would use ASCII transfer mode so that the file
created on the UNIX system would be in the UNIX text
format instead of the MS\-DOS text format.
.PP
You can retrieve an entire directory tree of files by
using the
.B -R
flag.
However, this will work only if the remote FTP server
is a UNIX server, or emulates UNIX's list output.
Example:
.RS 4
.sp
$ ncftpget -R ftp.probe.net /tmp /pub/ncftp
.br
.sp
.RE
This would create a /tmp/ncftp hierarchy.
.\"-------
.SH "DIAGNOSTICS"
.\"-------
.PP
.I ncftpget
returns the following exit values:
.TP 8
0
Success.
.TP 8
1
Could not connect to remote host.
.TP 8
2
Could not connect to remote host - timed out.
.TP 8
3
Transfer failed.
.TP 8
4
Transfer failed - timed out.
.TP 8
5
Directory change failed.
.TP 8
6
Directory change failed - timed out.
.TP 8
7
Malformed URL.
.TP 8
8
Usage error.
.TP 8
9
Error in login configuration file.
.TP 8
10
Library initialization failed.
.TP 8
11
Session initialization failed.
.\"-------
.SH "AUTHOR"
.\"-------
.PP
Mike Gleason, NcFTP Software (mgleason@ncftp.com).
.\"-------
.SH "SEE ALSO"
.\"-------
.PP
.IR ncftpput (1),
.IR ncftp (1),
.IR ftp (1),
.IR rcp (1),
.IR tftp (1).
.PP
.IR "LibNcFTP" " (http://www.ncftp.com/libncftp/)."


ChangeLog
2000-10-10  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_log.c: Removed my_vsnprintf() as it is now in
	string_funcs.c

	* verror.c: Removed my_vsnprintf() as it is now in
	string_funcs.c
	
	* string_funcs.c, string_funcs.h: Created these files.

	* myproxy_destroy.c (main): Bug fix - use created proxy file for
	authentication and not default proxy.
	(main): Bug fix - passphrase was not being correctly read.

2000-10-09  Von Welch  <vwelch@ncsa.uiuc.edu>

	* verror.c (my_vsnprintf): Fixed inverted logic in checking of
	HAVE_VSNPRINTF
	
2000-10-04  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_server.c (main): Print strerror() when failure occurs
	opening configuration file.

	* myproxy.c(parse_command): Fixed command_value type.

2000-10-03  Von Welch  <vwelch@ncsa.uiuc.edu>

	* ssl_utils.c: Fixes for openssl build with PEM callback
	differences (openssl has a user data parameter while ssleay does
	not).

	* ChangeLog: Created after 0.2alpha1 release



gridftp/gsincftp/source/sh_util/ncftpget.c
/* ncftpget.c
 *
 * A non-interactive utility to grab files from a remote FTP server.
 * Very useful in shell scripts!
 */

#include "syshdrs.h"

#if defined(WIN32) || defined(_WINDOWS)
#	include "..\ncftp\getopt.h"
#	define getopt Getopt
#	define optarg gOptArg
#	define optind gOptInd
	WSADATA wsaData;
	int wsaInit = 0;

	__inline void DisposeWinsock(int aUNUSED) { if (wsaInit > 0) WSACleanup(); wsaInit--; }
#	include "..\ncftp\util.h"
#	include "..\ncftp\spool.h"
#	include "..\ncftp\pref.h"
#else
#	include "../ncftp/util.h"
#	include "../ncftp/spool.h"
#	include "../ncftp/pref.h"
#endif

#include "gpshare.h"


FTPLibraryInfo gLib;
FTPConnectionInfo fi;

extern int gFirewallType;
extern char gFirewallHost[64];
extern char gFirewallUser[32];
extern char gFirewallPass[32];
extern unsigned int gFirewallPort;
extern char gFirewallExceptionList[256];
extern int gFwDataPortMode;
extern char gOS[], gVersion[];

extern char *optarg;
extern int optind;

static void
Usage(void)
{
	FILE *fp;

	fp = OpenPager();
	(void) fprintf(fp, "NcFTPGet %.5s\n\n", gVersion + 11);
	(void) fprintf(fp, "Usages:\n");
	(void) fprintf(fp, "  ncftpget [flags] remote-host local-dir remote-path-names...   (mode 1)\n");
	(void) fprintf(fp, "  ncftpget -f login.cfg [flags] local-dir remote-path-names...  (mode 2)\n");
	(void) fprintf(fp, "  ncftpget [flags] ftp://url.style.host/path/name               (mode 3)\n");
	(void) fprintf(fp, "\nFlags:\n\
  -u XX  Use username XX instead of anonymous.\n\
  -p XX  Use password XX with the username.\n\
  -P XX  Use port number XX instead of the default FTP service port (21).\n\
  -d XX  Use the file XX for debug logging.\n\
  -a     Use ASCII transfer type instead of binary.\n");
	(void) fprintf(fp, "\
  -t XX  Timeout after XX seconds.\n\
  -v/-V  Do (do not) use progress meters.\n\
  -f XX  Read the file XX for host, user, and password information.\n\
  -A     Append to local files, instead of overwriting them.\n");
	(void) fprintf(fp, "\
  -z/-Z  Do (do not) not try to resume downloads (default: -z).\n\
  -F     Use passive (PASV) data connections.\n\
  -DD    Delete remote file after successfully downloading it.\n\
  -b     Run in background (submit job to \"ncftpbatch\").\n\
  -B XX  Try setting the SO_RCVBUF size to XX.\n\
  -r XX  Redial XX times until connected.\n\
  -R     Recursive mode; copy whole directory trees.\n\
  -T     Do not try to use TAR mode with Recursive mode.\n");
	(void) fprintf(fp, "\nExamples:\n\
  ncftpget ftp.wustl.edu . /pub/README /pub/README.too\n\
  ncftpget ftp.wustl.edu . '/pub/README*'\n\
  ncftpget -R ftp.probe.net /tmp /pub/ncftpd  (ncftpd is a directory)\n\
  ncftpget ftp://ftp.wustl.edu/pub/README\n\
  ncftpget -u gleason -p my.password Bozo.probe.net . '/home/mjg/.*rc'\n\
  ncftpget -u gleason Bozo.probe.net . /home/mjg/foo.txt  (prompt for password)\n\
  ncftpget -f Bozo.cfg '/home/mjg/.*rc'\n\
  ncftpget -a -d /tmp/debug.log -t 60 ftp.wustl.edu . '/pub/README*'\n");

	(void) fprintf(fp, "\nLibrary version: %s.\n", gLibNcFTPVersion + 5);
	(void) fprintf(fp, "\nThis is a freeware program by Mike Gleason (mgleason@ncftp.com).\n");
	(void) fprintf(fp, "This was built using LibNcFTP (http://www.ncftp.com/libncftp).\n");

	ClosePager(fp);
	DisposeWinsock(0);
	exit(kExitUsage);
}	/* Usage */




static void
Abort(int sigNumUNUSED)
{
	signal(SIGINT, Abort);

	/* Hopefully the I/O operation in progress
	 * will complete, and we'll abort before
	 * it starts a new block.
	 */
	fi.cancelXfer++;

	/* If the user appears to be getting impatient,
	 * restore the default signal handler so the
	 * next ^C abends the program.
	 */
	if (fi.cancelXfer >= 2)
		signal(SIGINT, SIG_DFL);
}	/* Abort */




static int 
Copy(FTPCIPtr cip, char *dstdir, const char ** volatile files, int rflag, int xtype, int resumeflag, int appendflag, int deleteflag, int tarflag)
{
	int i;
	int result;
	const char *file;
	int rc = 0;

	for (i=0; ; i++) {
		file = files[i];
		if (file == NULL)
			break;
		result = FTPGetFiles3(cip, file, dstdir, rflag, kGlobYes, xtype, resumeflag, appendflag, deleteflag, tarflag, NoConfirmResumeDownloadProc, 0);
		if (result != 0) {
			(void) fprintf(stderr, "ncftpget: file retrieval error: %s.\n", FTPStrError(result));
			rc = result;
		}
	}
	return (rc);
}	/* Copy */




int
main(int argc, char **argv)
{
	int result, c;
	volatile int rflag = 0;
	volatile int xtype = kTypeBinary;
	volatile int appendflag = kAppendNo;
	volatile int resumeflag = kResumeYes;
	volatile int deleteflag = kDeleteNo;
	volatile int tarflag = kTarYes;
	int progmeters;
	char *dstdir = NULL;
	const char **flist;
	const char *errstr;
	volatile ExitStatus es;
	char url[256];
	char urlfile[128];
	char urldir[256];
	int urlxtype;
	LineList cdlist;
	LinePtr lp;
	int rc;
	int nD = 0;
	int batchmode = 0;
	int spooled = 0;
	int i;
	char *urlfilep;
	const char *urldirp;
	char *password;

	InitWinsock();
	result = FTPInitLibrary(&gLib);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpget: init library error %d (%s).\n", result, FTPStrError(result));
		exit(kExitInitLibraryFailed);
	}
	result = FTPInitConnectionInfo(&gLib, &fi, kDefaultFTPBufSize);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpget: init connection info error %d (%s).\n", result, FTPStrError(result));
		exit(kExitInitConnInfoFailed);
	}

	if (gFwDataPortMode >= 0)
		fi.dataPortMode = gFwDataPortMode;
	fi.debugLog = NULL;
	fi.errLog = stderr;
	fi.xferTimeout = 60 * 60;
	fi.connTimeout = 30;
	fi.ctrlTimeout = 135;
	(void) STRNCPY(fi.user, "anonymous");
	fi.host[0] = '\0';
	progmeters = GetDefaultProgressMeterSetting();
	urlfile[0] = '\0';
	InitLineList(&cdlist);

	while ((c = getopt(argc, argv, "P:u:j:p:e:d:t:aRTr:vVf:ADzZFbB:")) > 0) switch(c) {
		case 'P':
			fi.port = atoi(optarg);	
			break;
		case 'u':
			(void) STRNCPY(fi.user, optarg);
			break;
		case 'j':
			(void) STRNCPY(fi.acct, optarg);
			break;
		case 'p':
			(void) STRNCPY(fi.pass, optarg);	/* Don't recommend doing this! */
			break;
		case 'e':
			if (strcmp(optarg, "stdout") == 0)
				fi.errLog = stdout;
			else if (optarg[0] == '-')
				fi.errLog = stdout;
			else if (strcmp(optarg, "stderr") == 0)
				fi.errLog = stderr;
			else
				fi.errLog = fopen(optarg, FOPEN_APPEND_TEXT);
			break;
		case 'D':
			/* Require two -D's in case they typo. */
			nD++;
			break;
		case 'd':
			if (strcmp(optarg, "stdout") == 0)
				fi.debugLog = stdout;
			else if (optarg[0] == '-')
				fi.debugLog = stdout;
			else if (strcmp(optarg, "stderr") == 0)
				fi.debugLog = stderr;
			else
				fi.debugLog = fopen(optarg, FOPEN_APPEND_TEXT);
			break;
		case 't':
			SetTimeouts(&fi, optarg);
			break;
		case 'a':
			xtype = kTypeAscii;
			break;
		case 'r':
			SetRedial(&fi, optarg);
			break;
		case 'R':
			rflag = 1;
			break;
		case 'T':
			tarflag = 0;
			break;
		case 'v':
			progmeters = 1;
			break;
		case 'V':
			progmeters = 0;
			break;
		case 'f':
			ReadConfigFile(optarg, &fi);
			break;
		case 'A':
			appendflag = kAppendYes;
			break;
		case 'z':
			resumeflag = kResumeYes;
			break;
		case 'Z':
			resumeflag = kResumeNo;
			break;
		case 'F':
			if (fi.dataPortMode == kPassiveMode)
				fi.dataPortMode = kSendPortMode;
			else
				fi.dataPortMode = kPassiveMode;
			break;
		case 'b':
			batchmode++;
			break;
		case 'B':
			fi.dataSocketRBufSize = (size_t) atol(optarg);	
			break;
		default:
			Usage();
	}
	if (optind > argc - 1)
		Usage();

	InitUserInfo();
	LoadFirewallPrefs(0);

	if (progmeters != 0)
		fi.progress = PrStatBar;

	if (fi.host[0] == '\0') {
		(void) STRNCPY(url, argv[optind]);
		rc = FTPDecodeURL(&fi, url, &cdlist, urlfile, sizeof(urlfile), (int *) &urlxtype, NULL);
		if (rc == kMalformedURL) {
			(void) fprintf(stderr, "Malformed URL: %s\n", url);
			exit(kExitMalformedURL);
		} else if (rc == kNotURL) {
			/* This is what should happen most of the time. */
			if (optind > argc - 3)
				Usage();
			(void) STRNCPY(fi.host, argv[optind]);
			dstdir = argv[optind + 1];
			flist = (const char ** volatile) argv + optind + 2;
		} else {
			/* URL okay */
			flist = NULL;
			if ((urlfile[0] == '\0') && (rflag == 0)) {
				/* It was obviously a directory, and they didn't say -R. */
				(void) fprintf(stderr, "ncftpget: Use -R if you want the whole directory tree.\n");
				es = kExitUsage;
				exit((int) es);
			}
			xtype = urlxtype;
		}
	} else {
		if (optind > argc - 2)
			Usage();
		dstdir = argv[optind + 0];
		flist = (const char ** volatile) argv + optind + 1;
	}

	if (strcmp(fi.user, "anonymous") && strcmp(fi.user, "ftp")) {
		if (fi.pass[0] == '\0') {
			password = GetPass2("Password: ");		
			if (password != NULL) {
				(void) STRNCPY(fi.pass, password);
				/* Don't leave cleartext password in memory. */
				(void) memset(password, 0, strlen(fi.pass));
			}
		}
	}

	if (MayUseFirewall(fi.host, gFirewallType, gFirewallExceptionList) != 0) {
		fi.firewallType = gFirewallType; 
		(void) STRNCPY(fi.firewallHost, gFirewallHost);
		(void) STRNCPY(fi.firewallUser, gFirewallUser);
		(void) STRNCPY(fi.firewallPass, gFirewallPass);
		fi.firewallPort = gFirewallPort;
	}

	if (nD >= 2)
		deleteflag = kDeleteYes;

	if (batchmode != 0) {
		if (flist == NULL) {
			/* URL mode */

			urldir[0] = '\0';
			for (lp = cdlist.first; lp != NULL; lp = lp->next) {
				if (urldir[0] != '\0')
					STRNCAT(urldir, "/");
				STRNCAT(urldir, lp->line);
			}

			rc = SpoolX(
				"get",
				urlfile, 	/* Remote file */
				urldir,		/* Remote CWD */
				urlfile, 	/* Local file */
				".",		/* Local CWD */
				fi.host,
				fi.ip,
				fi.port,
				fi.user,
				fi.pass,
				xtype,
				rflag,
				deleteflag,
				fi.dataPortMode,
				(time_t) 0	/* when: now */
			);
			if (rc == 0) {
				fprintf(stdout, "  + Spooled; writing locally as %s/%s.\n", ".", urlfile);
				spooled++;
			}
		} else {
			/* List of files specified */
			for (i=0; flist[i] != NULL; i++) {
				STRNCPY(urlfile, flist[i]);
				urlfilep = strrchr(urlfile, '/');
				if (urlfilep == NULL) {
					urldirp = ".";
					urlfilep = urlfile;
				} else {
					urldirp = urlfile;
					*urlfilep++ = '\0';
				}

				rc = SpoolX(
					"get",
					urlfilep, 	/* Remote file */
					urldirp,	/* Remote CWD */
					urlfilep, 	/* Local file */
					dstdir,		/* Local CWD */
					fi.host,
					fi.ip,
					fi.port,
					fi.user,
					fi.pass,
					xtype,
					rflag,
					deleteflag,
					fi.dataPortMode,
					(time_t) 0	/* when: now */
				);
				if (rc == 0) {
					fprintf(stdout, "  + Spooled; writing locally as %s/%s.\n", urldirp, urlfilep);
					spooled++;
				}
			}
		}
		if (spooled > 0) {
			RunBatch(0, NULL);
			DisposeWinsock(0);
			exit(kExitSuccess);
		}
		DisposeWinsock(0);
		exit(kExitSpoolFailed);
	}
	
	es = kExitOpenTimedOut;
	errstr = "could not open remote host";
	if ((result = FTPOpenHost(&fi)) < 0) {
		(void) fprintf(stderr, "ncftpget: cannot open %s: %s.\n", fi.host, FTPStrError(result));
		es = kExitOpenFailed;
		DisposeWinsock(0);
		exit((int) es);
	}
	(void) FTPCmd(&fi, "CLNT NcFTPGet %.5s %s", gVersion + 11, gOS);
	if (flist == NULL) {
		/* URL mode */
		errstr = "could not change directory on remote host";
		es = kExitChdirTimedOut;
		for (lp = cdlist.first; lp != NULL; lp = lp->next) {
			if (FTPChdir(&fi, lp->line) != 0) {
				(void) fprintf(stderr, "ncftpget: cannot chdir to %s: %s.\n", lp->line, FTPStrError(fi.errNo));
				es = kExitChdirFailed;
				DisposeWinsock(0);
				exit((int) es);
			}
		}
		
		errstr = "could not read file from remote host";
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (FTPGetFiles3(&fi, urlfile, ".", rflag, kGlobYes, xtype, resumeflag, appendflag, deleteflag, tarflag, NoConfirmResumeDownloadProc, 0) < 0) {
			(void) fprintf(stderr, "ncftpget: file retrieval error: %s.\n", FTPStrError(fi.errNo));
			es = kExitXferFailed;
		} else {
			es = kExitSuccess;
		}
	} else {
		errstr = "could not read file from remote host";
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (Copy(&fi, dstdir, flist, rflag, xtype, resumeflag, appendflag, deleteflag, tarflag) < 0)
			es = kExitXferFailed;
		else
			es = kExitSuccess;
	}
	
	
	(void) FTPCloseHost(&fi);
	DisposeWinsock(0);
	
	exit((int) es);
}	/* main */
